Patch gatk-launch to allow it to run with standalone packaged GATK jars

droazen opened the pull request:
-Will now look for both shadow and spark jars in the same directory
 as the gatk-launch script, and use them if found. Also checks
 BIN_PATH for jars.

-Environment variable overrides GATK_SHADOW_JAR and GATK_SPARK_JAR
 take precedence over everything

-Wrapper script is used if found and GATK_SHADOW_JAR is not set.

Resolves #1693

droazen commented:
For @lbergelson to review. Note that a separate PR is coming soon to address #1987

lbergelson commented on gatk-launch:
```diff
@@ -105,36 +125,89 @@ def main(args):
+    return GATK_RUN_SCRIPT
+
+def getShadowJar(throwIfNotFound=True):
+    shadowJar = findJar("shadowJar.jar", envVariableOverride=GATK_SHADOW_JAR_ENV_VARIABLE)
```
Why is this `shadowJar.jar` and the spark jar is just `spark.jar`.  Should this just be `shadow.jar`?  Also, do we want to consider naming the shadow jar something else, like `complete.jar` or just `gatk.jar` vs `gatk-spark.jar`?

lbergelson replied:
Could you update build.gradle to fix this annoying inconsistency?

droazen replied:
It's `shadowJar.jar` because that's the suffix that gradle uses for our shadow jar -- I could attempt to re-configure `build.gradle` to use different names, though I think if we did that we'd have to rename `gradle shadowJar` to something else to be consistent.

lbergelson commented:
@droazen one comment.  I think we should consider the naming of our jars in the final packages though.  I think a folder with `gatk.jar` and `gatk-for-spark.jar` is clearer to users than `gatk-shadow.jar` and `gatk-spark.jar`

lbergelson commented:
Also, we might want to write some tests for gatk-launch, since it's starting to have lots of different invocation configurations and I find that any line of python code I don't actually run tends to have some horrible error that a compiler would have complained about.

droazen commented:
@lbergelson Addressed your comments, and also updated the README. There's now a `./gradlew localJar` that produces a shadow jar ending in `local.jar`. Please re-review.

lbergelson commented on README.md:
```diff
@@ -43,22 +43,30 @@ If you are looking for the codebase of the current production version of GATK, p
-      the versions of Spark and Hadoop installed on your cluster to be used.
+        ./gradlew localJar
+        
+    * The resulting jar will be in `build/libs` with a name like `gatk-pkg-VERSION-local.jar`
```
pkg?

droazen replied:
pkg -> package

lbergelson commented on README.md:
```diff
@@ -70,6 +78,8 @@ If you are looking for the codebase of the current production version of GATK, p
     * Requires Python 2.6 or greater.
     * You need to have built the GATK as described in the "Building GATK4" section above before running this script.
+    * Must be run either from the root of your git clone, or be located within the same directory as fully-packaged GATK jars
+      produced by `./gradlew localJar` and `./gradlew sparkJar`
```
Do you want to mention the environment variables here?

droazen replied:
Done

lbergelson commented:
@droazen Looks good to me.  Not sure how I feel about the pkg in the jar name.  It might seem a bit mysterious to people.  Packaged is a bit long though.  I might just leave it off?

droazen commented:
@lbergelson I think we need something in the name to distinguish it from the mini-jar produced by `gradle installDist` -- I will just spell out `package` for now

lbergelson commented:
maybe add a `d` to the end there?

droazen commented:
I like `gatk-package`, as in "this thing is the gatk package" -- `packaged` implies a mysterious unknown `packager`, which might disturb some users.

lbergelson commented:
👍
